### PR TITLE
Support import in DAML REPL

### PR DESCRIPTION
**Pull request by aherrmann-da**
* Adds support for import declarations at the DAML REPL prompt.
* Adds a test-case for REPL imports.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 1]
```diff
@@ -154,12 +154,31 @@ topologicalSort lfPkgs = map toPkg $ topSort $ transposeG graph
...
 
-type ReplM a = Repl.HaskelineT (State.StateT ReplState IO) a
+type ReplM = Repl.HaskelineT (State.StateT ReplState IO)
+
+data ReplInput
+  = ReplStatement (Stmt GhcPs (LHsExpr GhcPs))
+  | ReplImport (ImportDecl GhcPs)
+
+parseReplInput :: String -> DynFlags -> Either Error ReplInput
+parseReplInput input dflags = swapEither $ do
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:169]: `Either a b` has a  `Semigroup` instance that should do what you want here and avoids the somewhat confusing `swapEither` thing. Alternatively (no pun intended), you can use https://hackage.haskell.org/package/semigroupoids-5.2.2/docs/Data-Functor-Alt.html#t:Alt which is a bit clearer than the `Semigroup` instance imho.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 2]
```diff
@@ -154,12 +154,31 @@ topologicalSort lfPkgs = map toPkg $ topSort $ transposeG graph
...
+    -- We use @Either ReplInput Error@ to short-circuit on success.
+    -- The most common input will be statements. So, we try parsing statements
+    -- first and emit statement parse errors on failure.
+    e <- tryParse (parseStatement input dflags) (ReplStatement . unLoc)
+    _ <- tryParse (parseImport input dflags) (ReplImport . unLoc)
+    pure e
+  where
+    swapEither = either Right Left
+    tryParse :: ParseResult a -> (a -> ReplInput) -> Either ReplInput Error
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:178]: Slightly disappointed that neither `ParseResult` nor `P` have alternative instances or anything close to it.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 3]
```diff
@@ -191,54 +210,73 @@ runRepl opts mainDar replClient ideState = do
...
+        -> Stmt GhcPs (LHsExpr GhcPs)
+        -> ExceptT Error ReplM ()
+    handleStmt dflags line stmt = do
+        ReplState {imports, bindings, lineNumber} <- State.get
+        (bind, expr) <- maybe (throwError (UnsupportedStatement line)) pure (splitStmt stmt)
+        liftIO $ writeFileUTF8 (fromNormalizedFilePath $ lineFilePath lineNumber)
+            (renderModule dflags imports lineNumber bindings bind expr)
         -- Useful for debugging, probably best to put it behind a --debug flag
-        -- rendered <- liftIO  $readFileUTF8 (fromNormalizedFilePath $ lineFilePath i)
+        -- rendered <- liftIO $ readFileUTF8 (fromNormalizedFilePath $ lineFilePath lineNumber)
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:224]: We should probably just add that debugging flag :slightly_smiling_face:  (separately)

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 4]
```diff
@@ -191,54 +210,73 @@ runRepl opts mainDar replClient ideState = do
...
         -- Type of the statement so we can give it a type annotation
         -- and avoid incurring a typeclass constraint.
         stmtTy <- maybe (throwError TypeError) pure (exprTy $ tm_typechecked_source tcMod)
-        scriptRes <- liftIO $ ReplClient.runScript replClient (optDamlLfVersion opts) lfMod
-        case scriptRes of
-            Right _ -> pure (bind, stmtTy)
-            Left err -> throwError (ScriptError err)
+        -- If we get an error we don’t increment lineNumber and we
+        -- do not get a new binding
+        withExceptT ScriptError $ ExceptT $ liftIO $
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:236]: The `liftIO` here is unnecessary here I believe and confused me for a moment (I thought we were lifting into `ExceptT` and didn’t understand why we need the `withExceptT` then).

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 5]
```diff
@@ -154,12 +154,31 @@ topologicalSort lfPkgs = map toPkg $ topSort $ transposeG graph
     toPkg = (\(pkg, _, _) -> pkg) . fromVertex
 
 data ReplState = ReplState
-  { moduleNames :: [LF.ModuleName]
+  { imports :: [ImportDecl GhcPs]
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:157]: You probably want bang patterns on this given that you’re using `$!` below.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 6]
```diff
@@ -191,54 +210,73 @@ runRepl opts mainDar replClient ideState = do
...
+          , lineNumber = lineNumber + 1
+          }
+    handleImport
+        :: DynFlags
+        -> ImportDecl GhcPs
+        -> ExceptT Error ReplM ()
+    handleImport dflags imp = do
+        ReplState {imports, lineNumber} <- State.get
+        -- TODO[AH] Deduplicate imports.
+        let newImports = imp : imports
```
Comment (cocreature) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:251]: Hm, this seems surprisingly tricky with qualified import, explicit imports, …. I wonder if we can biggypack on `-ddump-minimal-imports` (no idea how and that drops too much).

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 1 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:169]: Thanks, I was sure there was one, but ended up looking at the wrong [`Alt`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Alt).

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 2 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:178]: Yeah, my thoughts exactly :disappointed:

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 4 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:236]: No, it's needed to lift into `ReplM`. The monad here is `ExceptT Error ReplM`. `ReplM` is used for state in `handleStmt`. See https://github.com/digital-asset/daml/pull/5386/commits/a78a847c17cfee1e5e4bdca14bde09469b14f9a7 and https://github.com/digital-asset/daml/pull/5386/commits/9d05b82077272101f9ddcbf7dc1fbf8621fdf43b .

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 5 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs]: Good catch, done.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 6 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:251]: Yeah, I haven't looked at all the options, yet. Duplicate imports don't make it fail, but only add bloat, so this didn't seem too critical.

Hmm, AFAIK `-ddump-minimal-imports` takes into account which symbols you actually use in the module. In the REPL use case that's almost always going to lead to too few imports. But, maybe we can re-use some piece of its implementation.

Another very dumb, but simple, solution could be to deduplicate on how the import declarations render. That doesn't yield a minimal set, but at least eliminates obvious duplicates.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 1 above)
Comment (aherrmann-da) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs]: Done. The only downside of `Alt` vs the swapped `Either` is that we have to attempt statement parsing last to get statement parse errors on invalid inputs. Shouldn't be a big issue though.

**Pull request merged**